GcsRestoreAcceptanceIT#shouldRunRestore flaky

**Summary**

> Try to answer the following as best as possible

- How often does the test fail? [once](https://github.com/camunda/zeebe/actions/runs/5197117029/jobs/9371563075?pr=12991) 
- Does it block your work? maybe
- Do we suspect that it is a real failure? not sure

**Failures**

> Outline known failure cases, e.g. a failed assertion and its stacktrace obtained from Jenkins

<details><summary>Example assertion failure</summary>
<pre>
Error:  Tests run: 2, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 85.94 s <<< FAILURE! -- in io.camunda.zeebe.it.backup.GcsRestoreAcceptanceIT
Error:  io.camunda.zeebe.it.backup.GcsRestoreAcceptanceIT.shouldRunRestore -- Time elapsed: 31.13 s <<< ERROR!
org.testcontainers.containers.ContainerLaunchException: Container startup failed for image localhost:5000/camunda/zeebe:current-test
	at org.testcontainers.containers.GenericContainer.doStart(GenericContainer.java:349)
	at org.testcontainers.containers.GenericContainer.start(GenericContainer.java:322)
	at org.testcontainers.junit.jupiter.TestcontainersExtension$StoreAdapter.start(TestcontainersExtension.java:274)
	at org.testcontainers.junit.jupiter.TestcontainersExtension$StoreAdapter.access$200(TestcontainersExtension.java:261)
	at org.testcontainers.junit.jupiter.TestcontainersExtension.lambda$null$3(TestcontainersExtension.java:76)
	at org.junit.jupiter.engine.execution.ExtensionValuesStore.lambda$getOrComputeIfAbsent$4(ExtensionValuesStore.java:86)
	at org.junit.jupiter.engine.execution.ExtensionValuesStore$MemoizingSupplier.computeValue(ExtensionValuesStore.java:[223](https://github.com/camunda/zeebe/actions/runs/5197117029/jobs/9371563075?pr=12991#step:10:224))
	at org.junit.jupiter.engine.execution.ExtensionValuesStore$MemoizingSupplier.get(ExtensionValuesStore.java:211)
	at org.junit.jupiter.engine.execution.ExtensionValuesStore$StoredValue.evaluate(ExtensionValuesStore.java:191)
	at org.junit.jupiter.engine.execution.ExtensionValuesStore$StoredValue.access$100(ExtensionValuesStore.java:171)
	at org.junit.jupiter.engine.execution.ExtensionValuesStore.getOrComputeIfAbsent(ExtensionValuesStore.java:89)
	at org.junit.jupiter.engine.execution.NamespaceAwareStore.getOrComputeIfAbsent(NamespaceAwareStore.java:53)
	at org.testcontainers.junit.jupiter.TestcontainersExtension.lambda$startContainers$4(TestcontainersExtension.java:76)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at org.testcontainers.junit.jupiter.TestcontainersExtension.startContainers(TestcontainersExtension.java:76)
	at org.testcontainers.junit.jupiter.TestcontainersExtension.findTestLifecycleAwareContainers(TestcontainersExtension.java:109)
	at org.testcontainers.junit.jupiter.TestcontainersExtension.beforeEach(TestcontainersExtension.java:94)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeBeforeEachCallbacks$2(TestMethodTestDescriptor.java:166)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeBeforeMethodsOrCallbacksUntilExceptionOccurs$6(TestMethodTestDescriptor.java:202)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeBeforeMethodsOrCallbacksUntilExceptionOccurs(TestMethodTestDescriptor.java:202)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeBeforeEachCallbacks(TestMethodTestDescriptor.java:165)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:132)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:68)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:151)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at org.junit.platform.engine.support.hierarchical.ForkJoinPoolHierarchicalTestExecutorService$ExclusiveTask.compute(ForkJoinPoolHierarchicalTestExecutorService.java:202)
	at org.junit.platform.engine.support.hierarchical.ForkJoinPoolHierarchicalTestExecutorService.executeNonConcurrentTasks(ForkJoinPoolHierarchicalTestExecutorService.java:172)
	at org.junit.platform.engine.support.hierarchical.ForkJoinPoolHierarchicalTestExecutorService.invokeAll(ForkJoinPoolHierarchicalTestExecutorService.java:152)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at org.junit.platform.engine.support.hierarchical.ForkJoinPoolHierarchicalTestExecutorService$ExclusiveTask.compute(ForkJoinPoolHierarchicalTestExecutorService.java:202)
	at org.junit.platform.engine.support.hierarchical.ForkJoinPoolHierarchicalTestExecutorService.invokeAll(ForkJoinPoolHierarchicalTestExecutorService.java:146)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at org.junit.platform.engine.support.hierarchical.ForkJoinPoolHierarchicalTestExecutorService$ExclusiveTask.compute(ForkJoinPoolHierarchicalTestExecutorService.java:202)
	at java.base/java.util.concurrent.RecursiveAction.exec(RecursiveAction.java:194)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1182)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1655)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1622)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165)
Caused by: org.rnorth.ducttape.RetryCountExceededException: Retry limit hit with exception
	at org.rnorth.ducttape.unreliables.Unreliables.retryUntilSuccess(Unreliables.java:88)
	at org.testcontainers.containers.GenericContainer.doStart(GenericContainer.java:334)
	... 59 more
Caused by: org.testcontainers.containers.ContainerLaunchException: Could not create/start container
	at org.testcontainers.containers.GenericContainer.tryStart(GenericContainer.java:553)
	at org.testcontainers.containers.GenericContainer.lambda$doStart$0(GenericContainer.java:344)
	at org.rnorth.ducttape.unreliables.Unreliables.retryUntilSuccess(Unreliables.java:81)
	... 60 more
Caused by: org.rnorth.ducttape.TimeoutException: java.util.concurrent.TimeoutException
	at org.rnorth.ducttape.timeouts.Timeouts.callFuture(Timeouts.java:70)
	at org.rnorth.ducttape.timeouts.Timeouts.doWithTimeout(Timeouts.java:60)
	at org.testcontainers.containers.wait.strategy.WaitAllStrategy.waitUntilReady(WaitAllStrategy.java:54)
	at org.testcontainers.containers.GenericContainer.waitUntilContainerStarted(GenericContainer.java:964)
	at org.testcontainers.containers.GenericContainer.tryStart(GenericContainer.java:490)
	... 62 more
Caused by: java.util.concurrent.TimeoutException
	at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:204)
	at org.rnorth.ducttape.timeouts.Timeouts.callFuture(Timeouts.java:65)
	... 66 more

Warning:  Flakes: 
Warning:  io.camunda.zeebe.it.backup.GcsRestoreAcceptanceIT.shouldRunRestore
Error:    Run 1: GcsRestoreAcceptanceIT.shouldRunRestore ? ContainerLaunch Container startup failed for image localhost:5000/camunda/zeebe:current-test
[INFO]   Run 2: PASS
</pre>
</details>

**Hypotheses**

> List any hypotheses if you have one; can be ommitted

**Logs**
[Test results for [IT] QA Integration Tests.zip](https://github.com/camunda/zeebe/files/11673860/Test.results.for.IT.QA.Integration.Tests.zip)

--- a/qa/integration-tests/src/test/java/io/camunda/zeebe/it/backup/GcsRestoreAcceptanceIT.java
+++ b/qa/integration-tests/src/test/java/io/camunda/zeebe/it/backup/GcsRestoreAcceptanceIT.java
@@ -16,7 +16,6 @@ import io.camunda.zeebe.backup.gcs.GcsBackupConfig;
 import io.camunda.zeebe.backup.gcs.GcsBackupStore;
 import io.camunda.zeebe.client.ZeebeClient;
 import io.camunda.zeebe.qa.util.actuator.BackupActuator;
-import io.camunda.zeebe.qa.util.testcontainers.ContainerLogsDumper;
 import io.camunda.zeebe.qa.util.testcontainers.GcsContainer;
 import io.camunda.zeebe.qa.util.testcontainers.ZeebeTestContainerDefaults;
 import io.camunda.zeebe.shared.management.openapi.models.BackupInfo;
@@ -24,58 +23,27 @@ import io.camunda.zeebe.shared.management.openapi.models.StateCode;
 import io.camunda.zeebe.shared.management.openapi.models.TakeBackupResponse;
 import io.zeebe.containers.ZeebeContainer;
 import java.time.Duration;
-import java.util.Map;
 import org.apache.commons.lang3.RandomStringUtils;
 import org.awaitility.Awaitility;
 import org.junit.jupiter.api.BeforeAll;
 import org.junit.jupiter.api.Test;
-import org.junit.jupiter.api.extension.RegisterExtension;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 import org.testcontainers.containers.ContainerLaunchException;
 import org.testcontainers.containers.GenericContainer;
 import org.testcontainers.containers.Network;
+import org.testcontainers.containers.output.Slf4jLogConsumer;
 import org.testcontainers.containers.startupcheck.OneShotStartupCheckStrategy;
 import org.testcontainers.junit.jupiter.Container;
 import org.testcontainers.junit.jupiter.Testcontainers;
 
 @Testcontainers
 final class GcsRestoreAcceptanceIT {
-
+  private static final Logger LOG = LoggerFactory.getLogger(GcsRestoreAcceptanceIT.class);
   private static final Network NETWORK = Network.newNetwork();
   private static final String BUCKET_NAME = RandomStringUtils.randomAlphabetic(10).toLowerCase();
 
-  private static final long BACKUP_ID = 1;
-
   @Container private static final GcsContainer GCS = new GcsContainer(NETWORK, "gcs.local");
-
-  @Container
-  private final ZeebeContainer zeebe =
-      new ZeebeContainer(ZeebeTestContainerDefaults.defaultTestImage())
-          .withNetwork(NETWORK)
-          .dependsOn(GCS)
-          .withoutTopologyCheck()
-          .withEnv("ZEEBE_BROKER_DATA_BACKUP_STORE", "GCS")
-          .withEnv("ZEEBE_BROKER_DATA_BACKUP_GCS_BUCKETNAME", BUCKET_NAME)
-          .withEnv("ZEEBE_BROKER_DATA_BACKUP_GCS_AUTH", "none")
-          .withEnv("ZEEBE_BROKER_DATA_BACKUP_GCS_HOST", GCS.internalEndpoint());
-
-  // No `@Container` annotation because we want to start this on demand
-  @SuppressWarnings("resource")
-  private final GenericContainer<?> restore =
-      new GenericContainer<>(ZeebeTestContainerDefaults.defaultTestImage())
-          .withNetwork(NETWORK)
-          .dependsOn(GCS)
-          .withStartupCheckStrategy(
-              new OneShotStartupCheckStrategy().withTimeout(Duration.ofMinutes(1)))
-          .withEnv("ZEEBE_RESTORE", "true")
-          .withEnv("ZEEBE_RESTORE_FROM_BACKUP_ID", Long.toString(BACKUP_ID))
-          .withEnv("ZEEBE_BROKER_DATA_BACKUP_STORE", "GCS")
-          .withEnv("ZEEBE_BROKER_DATA_BACKUP_GCS_BUCKETNAME", BUCKET_NAME)
-          .withEnv("ZEEBE_BROKER_DATA_BACKUP_GCS_AUTH", "none")
-          .withEnv("ZEEBE_BROKER_DATA_BACKUP_GCS_HOST", GCS.internalEndpoint());
-
-  @RegisterExtension
-  @SuppressWarnings("unused")
-  final ContainerLogsDumper logsWatcher = new ContainerLogsDumper(() -> Map.of("restore", restore));
 
   @BeforeAll
   static void setupBucket() throws Exception {
@@ -94,66 +62,76 @@ final class GcsRestoreAcceptanceIT {
   @Test
   void shouldRunRestore() {
     // given
-    final var actuator = BackupActuator.of(zeebe);
-    try (final var client =
-        ZeebeClient.newClientBuilder()
-            .gatewayAddress(zeebe.getExternalGatewayAddress())
-            .usePlaintext()
-            .build()) {
-      client.newPublishMessageCommand().messageName("name").correlationKey("key").send().join();
-    }
-    final var response = actuator.take(BACKUP_ID);
-    assertThat(response).isInstanceOf(TakeBackupResponse.class);
-    Awaitility.await("until a backup exists with the given ID")
-        .atMost(Duration.ofSeconds(60))
-        .ignoreExceptions() // 404 NOT_FOUND throws exception
-        .untilAsserted(
-            () -> {
-              final var status = actuator.status(BACKUP_ID);
-              assertThat(status)
-                  .extracting(BackupInfo::getBackupId, BackupInfo::getState)
-                  .containsExactly(1L, StateCode.COMPLETED);
-            });
+    final var backupId = 17;
+
+    // when
+    takeBackup(backupId);
 
     // then
-    assertThatNoException().isThrownBy(restore::start);
+    assertThatNoException().isThrownBy(() -> restoreBackup(backupId));
   }
 
   @Test
   void shouldFailForNonExistingBackup() {
-    // given
-    final var actuator = BackupActuator.of(zeebe);
-    try (final var client =
-        ZeebeClient.newClientBuilder()
-            .gatewayAddress(zeebe.getExternalGatewayAddress())
-            .usePlaintext()
-            .build()) {
-      client.newPublishMessageCommand().messageName("name").correlationKey("key").send().join();
-    }
-    final var response = actuator.take(BACKUP_ID);
-    assertThat(response).isInstanceOf(TakeBackupResponse.class);
-    Awaitility.await("until a backup exists with the given ID")
-        .atMost(Duration.ofSeconds(60))
-        .ignoreExceptions() // 404 NOT_FOUND throws exception
-        .untilAsserted(
-            () -> {
-              final var status = actuator.status(BACKUP_ID);
-              assertThat(status)
-                  .extracting(BackupInfo::getBackupId, BackupInfo::getState)
-                  .containsExactly(1L, StateCode.COMPLETED);
-            });
-
     // then -- restore container exits with an error code
     // we can't check the exit code directly, but we can observe that testcontainers was unable
     // to start the container.
-    assertThatExceptionOfType(ContainerLaunchException.class)
-        .isThrownBy(
-            () ->
-                restore
-                    .withStartupAttempts(1)
-                    .withEnv("ZEEBE_RESTORE_FROM_BACKUP_ID", "1234")
-                    .withStartupCheckStrategy(
-                        new OneShotStartupCheckStrategy().withTimeout(Duration.ofSeconds(10)))
-                    .start());
+    assertThatExceptionOfType(ContainerLaunchException.class).isThrownBy(() -> restoreBackup(1234));
+  }
+
+  private void takeBackup(final long backupId) {
+    try (final var zeebe =
+        new ZeebeContainer(ZeebeTestContainerDefaults.defaultTestImage())
+            .withLogConsumer(new Slf4jLogConsumer(LOG))
+            .withNetwork(NETWORK)
+            .dependsOn(GCS)
+            .withoutTopologyCheck()
+            .withEnv("ZEEBE_BROKER_DATA_BACKUP_STORE", "GCS")
+            .withEnv("ZEEBE_BROKER_DATA_BACKUP_GCS_BUCKETNAME", BUCKET_NAME)
+            .withEnv("ZEEBE_BROKER_DATA_BACKUP_GCS_AUTH", "none")
+            .withEnv("ZEEBE_BROKER_DATA_BACKUP_GCS_HOST", GCS.internalEndpoint())) {
+      zeebe.start();
+
+      final var actuator = BackupActuator.of(zeebe);
+
+      try (final var client =
+          ZeebeClient.newClientBuilder()
+              .gatewayAddress(zeebe.getExternalGatewayAddress())
+              .usePlaintext()
+              .build()) {
+        client.newPublishMessageCommand().messageName("name").correlationKey("key").send().join();
+      }
+
+      assertThat(actuator.take(backupId)).isInstanceOf(TakeBackupResponse.class);
+      Awaitility.await("until a backup exists with the given ID")
+          .atMost(Duration.ofSeconds(60))
+          .ignoreExceptions() // 404 NOT_FOUND throws exception
+          .untilAsserted(
+              () -> {
+                final var status = actuator.status(backupId);
+                assertThat(status)
+                    .extracting(BackupInfo::getBackupId, BackupInfo::getState)
+                    .containsExactly(backupId, StateCode.COMPLETED);
+              });
+    }
+  }
+
+  private void restoreBackup(final long backupId) {
+    try (final var restore =
+        new GenericContainer<>(ZeebeTestContainerDefaults.defaultTestImage())
+            .withLogConsumer(new Slf4jLogConsumer(LOG))
+            .withNetwork(NETWORK)
+            .dependsOn(GCS)
+            .withStartupAttempts(1)
+            .withStartupCheckStrategy(
+                new OneShotStartupCheckStrategy().withTimeout(Duration.ofMinutes(1)))
+            .withEnv("ZEEBE_RESTORE", "true")
+            .withEnv("ZEEBE_BROKER_DATA_BACKUP_STORE", "GCS")
+            .withEnv("ZEEBE_BROKER_DATA_BACKUP_GCS_BUCKETNAME", BUCKET_NAME)
+            .withEnv("ZEEBE_BROKER_DATA_BACKUP_GCS_AUTH", "none")
+            .withEnv("ZEEBE_BROKER_DATA_BACKUP_GCS_HOST", GCS.internalEndpoint())
+            .withEnv("ZEEBE_RESTORE_FROM_BACKUP_ID", String.valueOf(backupId))) {
+      restore.start();
+    }
   }
 }

